File: app/src/main/java/org/dark/keyboard/DarkIME.java
package org.dark.keyboard;

import android.inputmethodservice.InputMethodService;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.ExtractedTextRequest;

public class DarkIME extends InputMethodService implements LatinKeyboardBaseView.OnKeyboardActionListener {

    private LatinKeyboardView kv;
    private LatinKeyboard kbd;
    private boolean shiftState, ctrlActive, altActive, fnActive;

    @Override public void onCreate() {
        super.onCreate();
        Log.e("DK", "onCreate");
    }

    @Override public View onCreateInputView() {
        Log.e("DK", "onCreateInputView START");
        try {
            kbd = new LatinKeyboard(this, R.xml.kbd_qwerty, 0, 0.35f);
            kv = (LatinKeyboardView) getLayoutInflater().inflate(R.layout.input_ics, null)
                .findViewById(R.id.LatinkeyboardBaseView);
            kv.setKeyboard(kbd);
            kv.setOnKeyboardActionListener(this);
            kv.setLayoutParams(new ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));
            Log.e("DK", "onCreateInputView OK");
            return kv;
        } catch (Exception e) {
            Log.e("DK", "ERROR: " + e.getMessage(), e);
            android.widget.TextView tv = new android.widget.TextView(this);
            tv.setText("DK Error: " + e.getMessage());
            return tv;
        }
    }

    @Override public void onStartInputView(EditorInfo info, boolean restarting) {
        super.onStartInputView(info, restarting);
        Log.e("DK", "onStartInputView");
    }

    @Override public void onKey(int primaryCode, int[] keyCodes, int x, int y) {
        var ic = getCurrentInputConnection();
        if (ic == null) return;

        if (primaryCode == Keyboard.KEYCODE_SHIFT) {
            shiftState = !shiftState;
            kv.setShiftState(shiftState ? Keyboard.SHIFT_ON : Keyboard.SHIFT_OFF);
        } else if (primaryCode == Keyboard.KEYCODE_DELETE) {
            ic.deleteSurroundingText(1, 0);
        } else if (primaryCode == LatinKeyboardView.KEYCODE_CTRL_LEFT) {
            ctrlActive = !ctrlActive; kv.setCtrlIndicator(ctrlActive);
        } else if (primaryCode == LatinKeyboardView.KEYCODE_ALT_LEFT) {
            altActive = !altActive; kv.setAltIndicator(altActive);
        } else if (primaryCode == LatinKeyboardView.KEYCODE_FN) {
            fnActive = !fnActive;
            kbd = new LatinKeyboard(this, fnActive ? R.xml.kbd_symbols : R.xml.kbd_qwerty, 0, 0.35f);
            kv.setKeyboard(kbd);
        } else if (primaryCode == '\t') {
            ic.commitText("\t", 1);
        } else if (primaryCode == '\n') {
            ic.commitText("\n", 1);
            shiftState = false; kv.setShiftState(Keyboard.SHIFT_OFF);
        } else if (primaryCode == Keyboard.KEYCODE_MODE_CHANGE) {
            kbd = new LatinKeyboard(this, R.xml.kbd_symbols, 0, 0.35f);
            kv.setKeyboard(kbd);
        } else if (ctrlActive && primaryCode >= 'A' && primaryCode <= 'Z') {
            // Ctrl shortcuts
            switch (primaryCode) {
                case 'C': ic.performContextMenuAction(android.R.id.copy); break;
                case 'V': ic.performContextMenuAction(android.R.id.paste); break;
                case 'X': ic.performContextMenuAction(android.R.id.cut); break;
                case 'A': ic.performContextMenuAction(android.R.id.selectAll); break;
                default: ic.commitText(String.valueOf((char) primaryCode), 1);
            }
            ctrlActive = false; kv.setCtrlIndicator(false);
        } else {
            ic.commitText(String.valueOf((char) primaryCode), 1);
        }
    }

    @Override public void onRelease(int primaryCode) {}
    @Override public void onText(CharSequence text) {}
    @Override public boolean swipeRight() { return false; }
    @Override public boolean swipeLeft() { return false; }
    @Override public boolean swipeDown() { return false; }
    @Override public boolean swipeUp() { return false; }
    @Override public void onPress(int primaryCode) {}
    @Override public void onCancel() {}
}

File: app/src/main/java/org/dark/keyboard/DarkIME.kt
package org.dark.keyboard

import android.inputmethodservice.InputMethodService
import timber.log.Timber
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.LinearLayout

class DarkIME : InputMethodService(), SimpleKeyboardView.OnKeyListener {

    private var keyboardView: SimpleKeyboardView? = null

    override fun onCreate() {
        super.onCreate()
        Timber.d("onCreate")
    }

    override fun onCreateInputView(): View {
        Timber.d("onCreateInputView")
        try {
            val screenWidth = resources.displayMetrics.widthPixels
            val screenHeight = resources.displayMetrics.heightPixels

            val keyboard = SimpleKeyboard.fromXml(
                this,
                R.xml.kbd_qwerty,
                screenWidth,
                screenHeight
            )

            val container = LinearLayout(this).apply {
                orientation = LinearLayout.VERTICAL
                layoutParams = LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT
                )
            }

            keyboardView = SimpleKeyboardView(this).apply {
                setKeyboard(keyboard)
                onKeyListener = this@DarkIME
            }

            container.addView(keyboardView, LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT
            ))

            Timber.d("Keyboard created: ${keyboard.rows.size} rows, ${keyboard.allKeys.size} keys, h=${keyboard.totalHeight}")
            return container
        } catch (e: Exception) {
            Timber.e(e, "onCreateInputView error")
            return android.widget.TextView(this).apply { text = "Error: ${e.message}" }
        }
    }

    override fun onStartInputView(info: EditorInfo?, restarting: Boolean) {
        super.onStartInputView(info, restarting)
        Timber.d("onStartInputView")
    }

    override fun onKey(code: Int, shift: Boolean, ctrl: Boolean, alt: Boolean, fn: Boolean) {
        val ic = currentInputConnection ?: return

        when (code) {
            Key.CODE_DELETE -> {
                if (ctrl) {
                    ic.deleteSurroundingText(1, 0)
                } else {
                    ic.deleteSurroundingText(1, 0)
                }
            }
            Key.CODE_ENTER -> {
                val action = currentInputEditorInfo?.imeOptions ?: 0
                val maskedAction = action and EditorInfo.IME_MASK_ACTION
                if (maskedAction != EditorInfo.IME_ACTION_NONE) {
                    ic.performEditorAction(maskedAction)
                } else {
                    ic.commitText("\n", 1)
                }
            }
            Key.CODE_SPACE -> ic.commitText(" ", 1)
            Key.CODE_TAB -> ic.sendKeyEvent(
                android.view.KeyEvent(android.view.KeyEvent.ACTION_DOWN, android.view.KeyEvent.KEYCODE_TAB)
            )
            Key.CODE_SETTINGS -> {
                try {
                    val intent = android.content.Intent(android.provider.Settings.ACTION_INPUT_METHOD_SETTINGS)
                    intent.addFlags(android.content.Intent.FLAG_ACTIVITY_NEW_TASK)
                    startActivity(intent)
                } catch (e: Exception) {
                    Timber.e(e, "Cannot open settings")
                }
            }
            Key.CODE_F1 -> { }
            else -> {
                if (ctrl && code in 'A'.code..'Z'.code) {
                    when (code.toChar()) {
                        'A' -> ic.performContextMenuAction(android.R.id.selectAll)
                        'C' -> ic.performContextMenuAction(android.R.id.copy)
                        'V' -> ic.performContextMenuAction(android.R.id.paste)
                        'X' -> ic.performContextMenuAction(android.R.id.cut)
                        else -> ic.commitText(code.toChar().toString(), 1)
                    }
                } else {
                    val char = if (shift) code.toChar().uppercaseChar() else code.toChar()
                    ic.commitText(char.toString(), 1)
                }
            }
        }
    }

    override fun onText(text: CharSequence) {
        currentInputConnection?.commitText(text, 1)
    }
}